File: app/src/main/java/com/matchthree/game/model/Board.kt
package com.matchthree.game.model

/**
 * An immutable 9x9 (or config-sized) grid of gems. `null` marks an empty cell
 * (e.g. a cell awaiting a spawned gem). Engine passes return snapshots; the
 * grid itself is never mutated after construction.
 */
class Board private constructor(
    val width: Int,
    val height: Int,
    internal val cells: Array<Array<Gem?>>,
) {
    init {
        require(cells.size == height) { "cells rows != height" }
        require(cells.all { it.size == width }) { "cells cols != width" }
    }

    fun gemAt(row: Int, col: Int): Gem? =
        if (row in 0 until height && col in 0 until width) cells[row][col] else null

    fun gemAt(position: Position): Gem? = gemAt(position.row, position.col)

    fun isInside(position: Position): Boolean =
        position.row in 0 until height && position.col in 0 until width

    /** All board positions, row-major. */
    fun positions(): List<Position> =
        (0 until height).flatMap { row -> (0 until width).map { col -> Position(row, col) } }

    /** A deep-copied board with the two positions swapped. */
    fun withSwapped(a: Position, b: Position): Board {
        require(isInside(a) && isInside(b))
        val copy = copyCells()
        val gemA = copy[a.row][a.col]
        copy[a.row][a.col] = copy[b.row][b.col]
        copy[b.row][b.col] = gemA
        return Board(width, height, copy)
    }

    /** A deep-copied board with [gem] placed at [position]. */
    fun withGem(position: Position, gem: Gem?): Board {
        require(isInside(position))
        val copy = copyCells()
        copy[position.row][position.col] = gem
        return Board(width, height, copy)
    }

    private fun copyCells(): Array<Array<Gem?>> =
        Array(height) { row -> cells[row].copyOf() }

    companion object {
        /** Wraps a cell grid. The grid must match [width] x [height]. */
        fun of(width: Int, height: Int, cells: Array<Array<Gem?>>): Board =
            Board(width, height, cells)

        /** A board builder; [factory] returns the gem for each position (row-major). */
        fun create(width: Int, height: Int, factory: (Position) -> Gem?): Board {
            val cells = Array(height) { row ->
                Array(width) { col -> factory(Position(row, col)) }
            }
            return Board(width, height, cells)
        }
    }
}

File: Engine/Model/Board.cs
namespace MatchThree.Engine.Model;

public sealed class Board
{
    public int Width { get; }
    public int Height { get; }

    internal Gem?[,] Cells { get; }

    private Board(int width, int height, Gem?[,] cells)
    {
        if (cells.GetLength(0) != height)
            throw new ArgumentException($"cells rows != height ({cells.GetLength(0)} != {height})", nameof(cells));
        for (var row = 0; row < height; row++)
        {
            if (cells.GetLength(1) != width)
                throw new ArgumentException($"cells cols != width ({cells.GetLength(1)} != {width})", nameof(cells));
        }

        Width = width;
        Height = height;
        Cells = cells;
    }

    public Gem? GemAt(int row, int col) =>
        row >= 0 && row < Height && col >= 0 && col < Width ? Cells[row, col] : null;

    public Gem? GemAt(Position position) => GemAt(position.Row, position.Col);

    public Position? PositionOf(int gemId)
    {
        foreach (var p in Positions())
        {
            if (GemAt(p) is { } gem && gem.Id == gemId) return p;
        }
        return null;
    }

    public bool IsInside(Position position) =>
        position.Row >= 0 && position.Row < Height && position.Col >= 0 && position.Col < Width;

    public IEnumerable<Position> Positions()
    {
        for (var row = 0; row < Height; row++)
        {
            for (var col = 0; col < Width; col++)
            {
                yield return new Position(row, col);
            }
        }
    }

    public Board WithSwapped(Position a, Position b)
    {
        if (!IsInside(a) || !IsInside(b))
            throw new ArgumentException($"swap positions must be inside the board: {a}, {b}");

        var copy = CopyCells();
        var gemA = copy[a.Row, a.Col];
        copy[a.Row, a.Col] = copy[b.Row, b.Col];
        copy[b.Row, b.Col] = gemA;
        return new Board(Width, Height, copy);
    }

    public Board WithGem(Position position, Gem? gem)
    {
        if (!IsInside(position))
            throw new ArgumentException($"position must be inside the board: {position}");

        var copy = CopyCells();
        copy[position.Row, position.Col] = gem;
        return new Board(Width, Height, copy);
    }

    internal Gem?[,] CopyCells() => (Gem?[,])Cells.Clone();

    public static Board Of(int width, int height, Gem?[,] cells) => new(width, height, cells);

    public static Board Create(int width, int height, Func<Position, Gem?> factory)
    {
        var cells = new Gem?[height, width];
        for (var row = 0; row < height; row++)
        {
            for (var col = 0; col < width; col++)
            {
                cells[row, col] = factory(new Position(row, col));
            }
        }
        return new Board(width, height, cells);
    }
}